internal/operator/statefulset/builder.go

```diff
@@ -0,0 +1,871 @@
+					DefaultMode: pointer.Int32(common.OwnerReadWriteGroupReadPermissions),
+				},
+			},
+		},
+		{
+			Name: common.EtcdPeerServerTLSVolumeName,
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName:  peerTLSConfig.ServerTLSSecretRef.Name,
+					DefaultMode: pointer.Int32(common.OwnerReadWriteGroupReadPermissions),
+				},
+			},
+		},
+	}
+}
+
+func (b *stsBuilder) getBackupRestoreTLSVolumes() []corev1.Volume {
+	tlsConfig := b.etcd.Spec.Backup.TLS
+	return []corev1.Volume{
+		{
+			Name: common.BackupRestoreCAVolumeName,
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName:  tlsConfig.TLSCASecretRef.Name,
+					DefaultMode: pointer.Int32(common.OwnerReadWriteGroupReadPermissions),
+				},
+			},
+		},
+		{
+			Name: common.BackupRestoreServerTLSVolumeName,
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName:  tlsConfig.ServerTLSSecretRef.Name,
+					DefaultMode: pointer.Int32(common.OwnerReadWriteGroupReadPermissions),
+				},
+			},
+		},
+		{
+			Name: common.BackupRestoreClientTLSVolumeName,
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName:  tlsConfig.ClientTLSSecretRef.Name,
+					DefaultMode: pointer.Int32(common.OwnerReadWriteGroupReadPermissions),
+				},
+			},
+		},
+	}
+}
+
+func (b *stsBuilder) getBackupVolume(ctx component.OperatorContext) (*corev1.Volume, error) {
+	if b.provider == nil {
+		return nil, nil
+	}
+	store := b.etcd.Spec.Backup.Store
+	switch *b.provider {
+	case utils.Local:
+		hostPath, err := utils.GetHostMountPathFromSecretRef(ctx, b.client, b.logger, store, b.etcd.GetNamespace())
+		if err != nil {
+			return nil, fmt.Errorf("error getting host mount path for etcd: %v Err: %w", b.etcd.GetNamespaceName(), err)
+		}
+
+		hpt := corev1.HostPathDirectory
+		return &corev1.Volume{
+			Name: common.LocalBackupVolumeName,
+			VolumeSource: corev1.VolumeSource{
+				HostPath: &corev1.HostPathVolumeSource{
+					Path: hostPath + "/" + pointer.StringDeref(store.Container, ""),
+					Type: &hpt,
+				},
+			},
+		}, nil
+	case utils.GCS, utils.S3, utils.OSS, utils.ABS, utils.Swift, utils.OCS:
+		if store.SecretRef == nil {
+			return nil, fmt.Errorf("etcd: %v, no secretRef configured for backup store", b.etcd.GetNamespaceName())
+		}
+
+		return &corev1.Volume{
+			Name: common.ProviderBackupSecretVolumeName,
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName:  store.SecretRef.Name,
+					DefaultMode: pointer.Int32(common.OwnerReadWriteGroupReadPermissions),
+				},
+			},
+		}, nil
+	}
+	return nil, nil
+}
+
+func getBackupStoreProvider(etcd *druidv1alpha1.Etcd) (*string, error) {
+	if !etcd.IsBackupStoreEnabled() {
+		return nil, nil
+	}
+	provider, err := utils.StorageProviderFromInfraProvider(etcd.Spec.Backup.Store.Provider)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
There are other problems for callers of this function. You can see usages and spot them. Ideally the provider should always be provided but since this is a common function used at multiple places we have decided not to change it in this PR (only to not increase the scope of this PR). Please make a note of this and jot down all open points that we are not addressing as part of this PR and then create an issue out of it.

Reply:
Yes. `infra` can be `nil` for cases where backups are not enabled for the etcd, like `etcd-events` in gardener.

`infra` could possibly also be `""`, if somebody is creating an `Etcd` with a `spec.backup.store` section but not populating it, resulting in an empty string for field `provider`. Checking the API now, I see that while `spec.backup.store` itself is optional ([ref](https://github.com/gardener/etcd-druid/blob/4cb6e5cd5cdf92225c2c45770c790e3df090005c/api/v1alpha1/types_etcd.go#L145-L147)), I see that `spec.backup.store.provider` is also optional within that ([ref](https://github.com/gardener/etcd-druid/blob/4cb6e5cd5cdf92225c2c45770c790e3df090005c/api/v1alpha1/types_common.go#L23-L25)), meaning if no value is set for the provider, then ideally it must be `nil` and not an empty string, since it is a pointer to a string. So technically, we don't need the `len(*infra) == 0` check in `utils.StorageProviderFromInfraProvider()`, and should ideally return an error if at all the user has used a pointer to an empty string here, because we should consider that as an unknown provider.

I will remove the unnecessary part of the check in the function.